utils/upgrade

```diff
@@ -0,0 +1,345 @@
+
+        return Success()
+
+
+class InsertNewModule(InsertModule):
+    log = "Try to insert new caching module"
+
+
+class DryRun(InsertModule):
+    log = "Perform dry run to check upgrade data integrity"
+    options = {"installed": False, "dry_run": 1}
+
+    def do_work(self):
+        result = super().do_work()
+        if not isinstance(result, Success):
+            return Failure("Dry run failed")
+
+        try:
+            logging.info(
+                "Insert with dry_run set succeeded. Removing module to attempt final insertion."
+            )
+            remove_module("cas_cache")
+        except Exception as e:
+            return Failure("Could not unload module. Reason: {e}")
+
+        return result
+
+
+class InsertInstalledModule(InsertModule):
+    log = "Restore old cas_cache module"
+    module_path = "cas_cache"
+    options = {"installed": True}
+
+
+class PrepareForUpgrade(UpgradeState):
+    log = "Preparing Open CAS for upgrade"
+
+    def do_work(self):
+        try:
+            logging.info("Switching CAS to upgrade mode")
+            opencas.casadm.start_upgrade()
+        except opencas.casadm.CasadmError as e:
+            return Failure(e)
+
+        return Success()
+
+
+class RemoveModule(UpgradeState):
+    log = "Removing cas_cache module"
+
+    def do_work(self):
+        try:
+            logging.info("Removing cas_cache module")
+            remove_module("cas_cache")
+        except Exception as e:
+            return Failure(f"Failed to remove cas_cache module. Reason: {e}")
+
+        return Success()
+
+
+class SetSchedulersToNoop(UpgradeState):
+    log = "Setting core devices schedulers to noop"
+
+    def do_work(self):
+        unique_core_sysfs_devices = {
+            get_device_sysfs_path(core["device"])
+            for _, core in opencas.get_devices_state()["cores"].items()
+        }
+
+        self.state_machine.schedulers = {}
+        for core in unique_core_sysfs_devices:
+            current, available = get_device_schedulers(core)
+            self.state_machine.schedulers[core] = current
+
+            if current in ["noop", "none"]:
+                logging.info(f"Core {core} already uses {current} scheduler. Skipping.")
+            elif "noop" in available:
+                logging.info(f"Switching scheduler for {core}: {current} => noop")
+                set_device_scheduler(core, "noop")
+            elif "none" in available:
+                logging.info(f"Switching scheduler for {core}: {current} => none")
+                set_device_scheduler(core, "none")
+            else:
+                logging.info(f"No apropriate scheduler available for {core}. Skipping.")
+
+        return Success()
+
+
+class RestoreCoreSchedulers(UpgradeState):
+    log = "Restoring core devices schedulers"
+
+    def do_work(self):
+        for core, scheduler in self.state_machine.schedulers.items():
+            current, _ = get_device_schedulers(core)
```

Comment:
I'd avoid '_' where possible - get_device_schedulers(core)[0] gets the job done

Reply:
Fixed